### PR TITLE
[Project] Implementing Project CLI

### PR DESCRIPTION
**Pull request by simon-mo**
## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/projects/scripts.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    "--verbose", help="If set, print the validated file", is_flag=True)
+def validate(verbose):
+    try:
+        proj = ray.projects.load_project(os.getcwd())
+        print("🍰 Project files validated!", file=sys.stderr)
+        if verbose:
+            print(proj)
+    except (jsonschema.exceptions.ValidationError, ValueError) as e:
+        print("💔 Validation failed for the following reason", file=sys.stderr)
+        raise e
```
Comment (richardliaw) [inline python/ray/projects/scripts.py]: try raising click exceptions; you won't see the stack trace

File: python/ray/projects/scripts.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+# File layout for generated project files
+# user-dir/
+#   .rayproject/
+#     project.yaml
+#     cluster.yaml
+#     requirements.txt
+PROJECT_DIR = ".rayproject"
+PROJECT_YAML = os.path.join(PROJECT_DIR, "project.yaml")
+CLUSTER_YAML = os.path.join(PROJECT_DIR, "cluster.yaml")
+REQ_TXT = os.path.join(PROJECT_DIR, "requirements.txt")
```
Comment (pcmoritz) [inline python/ray/projects/scripts.py]: Let's go with REQUIREMENTS_TXT here to not be too cryptic :)

File: python/ray/projects/scripts.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import logging
+import os
+import sys
+from shutil import copyfile
```
Comment (pcmoritz) [inline python/ray/projects/scripts.py]: Is this available by default? Are we using it anywhere else?

File: python/ray/projects/schema.json [hunk 4]
```diff
@@ -60,5 +60,8 @@
       }
     }
   },
-  "required": ["name", "cluster"]
-}
+  "required": [
+    "name",
+    "cluster"
+  ]
+}
```
Comment (pcmoritz) [inline python/ray/projects/schema.json]: We should have trailing newlines (also in the other files, maybe you can configure your editor to add it)

File: python/ray/projects/scripts.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+# File layout for templates file
+# RAY/.../projects/
+#   templates/
+#     cluster_template.yaml
+#     project_template.yaml
+#     requirements.txt
+_THIS_FILE_DIR = os.path.split(os.path.abspath(__file__))[0]
+_TEMPLATE_DIR = os.path.join(_THIS_FILE_DIR, "templates")
+PROJECT_TMPL = os.path.join(_TEMPLATE_DIR, "project_template.yaml")
```
Comment (pcmoritz) [inline python/ray/projects/scripts.py]: I'd go with PROJECT_TEMPLATE here

File: python/setup.py [hunk 6]
```diff
@@ -43,6 +46,12 @@
     "ray/autoscaler/local/example-full.yaml",
 ]
 
+ray_project_files = [
```
Comment (pcmoritz) [inline python/setup.py]: nice!

File: python/ray/projects/templates/project_template.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+# repo: The URL of the repo this project is part of
+
+cluster: {{cluster}}
+
+environment:
+  # dockerfile: The dockerfile to be built and ran the commandas with
+  # dockerimage: The docker image to be used to run the project in. Like ubuntu:18.04
+  requirements: {{requirements}}
+
+  shell: # shell commands to be ran for environment setup.
```
Comment (pcmoritz) [inline python/ray/projects/templates/project_template.yaml]: # shell -> # Shell

File: python/ray/projects/templates/project_template.yaml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    - echo "Setting up the environment"
+
+commands:
+  - name: first command
```
Comment (pcmoritz) [inline python/ray/projects/templates/project_template.yaml]: command names shouldn't include spaces (replace with first-command?)

File: python/ray/projects/scripts.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+def project_cli():
+    pass
+
+
+@project_cli.command(help="Validate current project spec")
+@click.option(
+    "--verbose", help="If set, print the validated file", is_flag=True)
+def validate(verbose):
+    try:
+        proj = ray.projects.load_project(os.getcwd())
```
Comment (pcmoritz) [inline python/ray/projects/scripts.py]: proj -> project

File: python/ray/projects/scripts.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+    if requirements is None:
+        logging.warn("Using default requirements.txt")
+        # no templating required, just copy the file
+        copyfile(REQ_TXT_TMPL, REQ_TXT)
+
+        requirements = REQ_TXT
+
+    with open(PROJECT_TMPL) as f:
+        proj_tmpl = f.read()
```
Comment (pcmoritz) [inline python/ray/projects/scripts.py]: proj_tmpl -> project_template

File: python/ray/projects/scripts.py (hunk 3 above)
Comment (simon-mo) [reply python/ray/projects/scripts.py]: Yes. The following files use it:

```
python/setup.py
rllib/tests/test_checkpoint_restore.py
rllib/tests/test_io.py
python/ray/projects/scripts.py
python/ray/tests/test_basic.py
python/ray/tests/test_autoscaler.py
python/ray/tests/test_tempfile.py
python/ray/tune/trainable.py
python/ray/tune/tests/test_tune_restore.py
python/ray/tune/tests/test_trial_scheduler.py
python/ray/tune/tests/test_experiment_analysis.py
python/ray/tune/tests/test_trial_runner.py
python/ray/tune/tests/test_cluster.py
python/ray/tune/schedulers/pbt.py
```

I think we can safely assume it exists

File: python/ray/projects/scripts.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+        with open(CLUSTER_TMPL) as f:
+            tmpl = f.read().replace(r"{{name}}", project_name)
+        with open(CLUSTER_YAML, "w") as f:
+            f.write(tmpl)
+
+        cluster_yaml = CLUSTER_YAML
+
+    if requirements is None:
+        logging.warn("Using default requirements.txt")
```
Comment (richardliaw) [inline python/ray/projects/scripts.py]: `logger = logging.getLogger`?

File: python/ray/projects/scripts.py (hunk 10 above)
Comment (richardliaw) [reply python/ray/projects/scripts.py]: agree here - in general, would be good to be more verbose

File: python/ray/projects/scripts.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+#     project_template.yaml
+#     requirements.txt
+_THIS_FILE_DIR = os.path.split(os.path.abspath(__file__))[0]
+_TEMPLATE_DIR = os.path.join(_THIS_FILE_DIR, "templates")
+PROJECT_TMPL = os.path.join(_TEMPLATE_DIR, "project_template.yaml")
+CLUSTER_TMPL = os.path.join(_TEMPLATE_DIR, "cluster_template.yaml")
+REQ_TXT_TMPL = os.path.join(_TEMPLATE_DIR, "requirements.txt")
+
+
+@click.group("project", help="Commands working with ray project")
```
Comment (richardliaw) [inline python/ray/projects/scripts.py]: this needs a better help string. @pcmoritz?

File: python/ray/projects/scripts.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
+import logging
```
Comment (richardliaw) [inline python/ray/projects/scripts.py]: py2to3

File: python/ray/projects/scripts.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+CLUSTER_TMPL = os.path.join(_TEMPLATE_DIR, "cluster_template.yaml")
+REQ_TXT_TMPL = os.path.join(_TEMPLATE_DIR, "requirements.txt")
+
+
+@click.group("project", help="Commands working with ray project")
+def project_cli():
+    pass
+
+
+@project_cli.command(help="Validate current project spec")
```
Comment (richardliaw) [inline python/ray/projects/scripts.py]: Don't you need to point `setup.py`to setup the cli for this?

Comment (richardliaw) [reply python/ray/projects/scripts.py]: nevermind i see the import

File: python/ray/projects/scripts.py (hunk 12 above)
Comment (richardliaw) [reply python/ray/projects/scripts.py]: Actually, can you mark this explicitly as experimental?

**Review by richardliaw — COMMENTED** [review]
Can you add a page to the docs for this marking it as experimental?

**Comment by simon-mo** [comment]
@richardliaw I believe @pcmoritz is adding docs in following up PR. 

Experimental is marked here:
```
Usage: ray [OPTIONS] COMMAND [ARGS]...

Options:
  --logging-level TEXT   The logging level threshold, choices=['debug',
                         'info', 'warning', 'error', 'critical'],
                         default='info'
  --logging-format TEXT  The logging format. default='%(asctime)s
                         %(levelname)s %(filename)s:%(lineno)s -- %(message)s'
  --help                 Show this message and exit.

Commands:
...
  project           [Experimental] Commands working with ray project
  rsync-down
  rsync-up
  rsync_down
  rsync_up
  session           [Experimental] Commands working with ray session
...
```

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**